AmplifyPlugins/Core/AWSPluginsCore/Model/GraphQLRequest/GraphQLRequest+Model.swift

```diff
@@ -8,18 +8,123 @@
 import Amplify
 import Foundation
 
-/// Extension that provides an integration layer between `Model`, `GraphQLDocument` and `GraphQLRequest`.
-/// This is particularly useful when using the GraphQL API to interact with static types that conform to
-/// the `Model` protocol.
-extension GraphQLRequest {
+// MARK: - Protocol
+
+/// Protocol that represents the integration between `GraphQLRequest` and `Model`.
+///
+/// The methods defined here are used to build a valid `GraphQLRequest` from types
+/// conforming to `Model`.
+protocol ModelGraphQLRequestFactory {
+
+    // MARK: Query
+
+    /// Creates a `GraphQLRequest` that represents a query that expects multiple values as a result.
+    /// The request will be created with the correct document based on the `ModelSchema` and
+    /// variables based on the the predicate.
+    ///
+    /// - Parameters:
+    ///   - modelType: the metatype of the model
+    ///   - predicate: an optional predicate containing the criteria for the query
+    /// - Returns: a valid `GraphQLRequest` instance
+    ///
+    /// - seealso: `GraphQLQuery`, `GraphQLQueryType.list`
+    static func list<M: Model>(_ modelType: M.Type,
+                               where predicate: QueryPredicate?) -> GraphQLRequest<[M]>
+
+    /// Creates a `GraphQLRequest` that represents a query that expects a single value as a result.
+    /// The request will be created with the correct correct document based on the `ModelSchema` and
+    /// variables based on given `id`.
+    ///
+    /// - Parameters:
+    ///   - modelType: the metatype of the model
+    ///   - id: the model identifier
+    /// - Returns: a valid `GraphQLRequest` instance
+    ///
+    /// - seealso: `GraphQLQuery`, `GraphQLQueryType.get`
+    static func get<M: Model>(_ modelType: M.Type, byId id: String) -> GraphQLRequest<M?>
+
+    // MARK: Mutation
 
     /// Creates a `GraphQLRequest` that represents a mutation of a given `type` for a `model` instance.
     ///
     /// - Parameters:
     ///   - model: the model instance populated with values
     ///   - predicate: a predicate passed as the condition to apply the mutation
     ///   - type: the mutation type, either `.create`, `.update`, or `.delete`
-    /// - Returns: the `GraphQLRequest` ready to be used
+    /// - Returns: a valid `GraphQLRequest` instance
+    static func mutation<M: Model>(of model: M,
+                                   where predicate: QueryPredicate?,
+                                   type: GraphQLMutationType) -> GraphQLRequest<M>
+
```

Comment:
Do we need this if we're exposing create/update/delete individual factory methods?

Reply:
This is useful when creating a mutation based on some logic (i.e. should it be a `create` or `update`?). We use this internally in the sync engine, so I kept it to reduce the impact of the changes.